lua/ttt2/libraries/modelbuilder.lua

```diff
@@ -0,0 +1,316 @@
+---@field material string The model material
+---@field skin number The skin index of the model
+---@field bodygroup table A table of bodygroups
+
+---@class BoneData
+---@field scale Vector The scale multiplier
+---@field pos Vector The position offset
+---@field angle Angle The rotation offset
+
+if SERVER then
+	AddCSLuaFile()
+
+	return
+end
+
+modelbuilder = {}
+
+local propertiesMaterial = {"nocull", "additive", "vertexalpha", "vertexcolor", "ignorez"}
+
+---
+-- Creates a clientside model that is used as a view or world model with the provided model data.
+-- @param Entity wep The weapon for which the model should be created
+-- @param ModelData modelData The model data for the model
+-- @realm client
+function modelbuilder.CreateModel(wep, modelData)
+	-- handle a model being added to the view or world model
+	if modelData.type == "Model" and modelData.model and modelData.model ~= ""
+		and (not IsValid(modelData.modelEnt) or modelData.createdModel ~= modelData.model)
+		and string.find(modelData.model, ".mdl") and file.Exists (modelData.model, "GAME")
+	then
+		modelData.modelEnt = ClientsideModel(modelData.model, RENDER_GROUP_VIEW_MODEL_OPAQUE) --todo check correct render group
+
+		if IsValid(modelData.modelEnt) then
+			modelData.modelEnt:SetPos(wep:GetPos())
+			modelData.modelEnt:SetAngles(wep:GetAngles())
+			modelData.modelEnt:SetParent(wep)
+			modelData.modelEnt:SetNoDraw(true)
+			modelData.createdModel = modelData.model
+		else
+			modelData.modelEnt = nil
+		end
+
+	-- handle a sprite being added to the view or world model
+	elseif modelData.type == "Sprite" and modelData.sprite and modelData.sprite ~= ""
+		and (not modelData.spriteMaterial or modelData.createdSprite ~= modelData.sprite)
+		and file.Exists ("materials/" .. modelData.sprite .. ".vmt", "GAME")
+	then
+		local name = modelData.sprite .. "-"
+		local params = {
+			["$basetexture"] = modelData.sprite
+		}
+
+		-- make sure we create a unique name based on the selected options
+		for i = 1, #propertiesMaterial do
+			local property = propertiesMaterial[i]
+
+			if modelData[property] then
+				params["$" .. property] = 1
+				name = name .. "1"
+			else
+				name = name .. "0"
+			end
+		end
+
+		modelData.createdSprite = modelData.sprite
+		modelData.spriteMaterial = CreateMaterial(name, "UnlitGeneric", params)
+	end
+end
+
+---
+-- Resets the bone positions of a view model.
+-- @param Entity viewModel The view model of the weapon
+-- @realm client
+function modelbuilder.ResetBonePositions(viewModel)
+	local boneCount = viewModel:GetBoneCount()
+
+	if not boneCount then return end
+
+	for i = 0, boneCount do
+		viewModel:ManipulateBoneScale(i, Vector(1, 1, 1))
+		viewModel:ManipulateBoneAngles(i, Angle(0, 0, 0))
+		viewModel:ManipulateBonePosition(i, Vector(0, 0, 0))
+	end
+end
+
+---
+-- Updates the bone positions of the view model of a weapon. Is used to hide a
+-- default view model of a weapon.
```

Comment:
How is this used to hide the default view model?

And please explain what "update" means in this context. Like update the bone positions to what?

Reply:
Maybe I should remove this then. This is taken from the old addon and how they tried to hide the weapon in the view model. But I found a better way that I'm using now. Therefore this is no longer needed. However I thought I keep it since I already put the work in it and it might be useful for something else as well, I'm not sure

Reply:
I think this can also be used to make more advanced stuff, like custom finger positions etc for addons such as the harpoon. So I will keep it, but I'm not knowledgeable enough yet to give this a good documentation, sorry

Reply:
Just describe what the code below tells you it does. Like literally that this applies custom bone modifications on the weapon. And if none are there it will just reset the bone positions.